src/main/java/com/qouteall/immersive_portals/Helper.java

```diff
@@ -658,4 +679,279 @@ public static Quaternion interpolateQuaternionNaive(
+        });
+
+        return hits;
+    }
+
+    /**
+     * @see #withSwitchedContext(World, Supplier) 
+     */
+    @Environment(EnvType.CLIENT)
+    private static <T> T withSwitchedContextClient(ClientWorld world, Supplier<T> func) {
+        boolean wasSwitched = BlockManipulationClient.isContextSwitched;
+        BlockManipulationClient.isContextSwitched = true;
+
+        MinecraftClient mc = MinecraftClient.getInstance();
+        ClientWorld lastWorld = mc.world;
+        mc.world = world;
+
+        try {
+            return func.get();
+        } finally {
+            mc.world = lastWorld;
+            BlockManipulationClient.isContextSwitched = wasSwitched;
+        }
+    }
+
+    /**
+     * @see #withSwitchedContext(World, Supplier)
+     */
+    @SuppressWarnings("unused")
+    private static <T> T withSwitchedContextServer(ServerWorld world, Supplier<T> func) {
+        // lol
+        return func.get();
+    }
+
+    /**
+     * Execute {@code func} with the world being set to {@code world}, hopefully bypassing any issues that may be
+     * related to mutating a world that is not currently set as the current world.
+     * <p>
+     * You may safely nest this function within other context switches. It works on both the client and the server.
+     *
+     * @param world The world to switch the context to. The context will be restored when {@code func} is complete.
+     * @param func  The function to execute while the context is switched.
+     * @param <T>   The return type of {@code func}.
+     * @return Whatever {@code func} returned.
+     */
+    private static <T> T withSwitchedContext(World world, Supplier<T> func) {
+        if (world.isClient) {
+            return withSwitchedContextClient((ClientWorld) world, func);
+        } else {
+            return withSwitchedContextServer((ServerWorld) world, func);
+        }
+    }
+
+    /**
+     * @see Helper#rayTrace(World, RayTraceContext, boolean)
+     * @author LoganDark
+     */
+    private static Pair<BlockHitResult, List<Portal>> rayTrace(
+        World world,
+        RayTraceContext context,
+        boolean includeGlobalPortals,
+        List<Portal> portals
+    ) {
+        Vec3d start = context.getStart();
+        Vec3d end = context.getEnd();
+
+        // If we're past the max portal layer, don't let the player target behind this portal, create a missed result
+        if (portals.size() > Global.maxPortalLayer) {
+            Vec3d diff = end.subtract(start);
+
+            return new Pair<>(
+                BlockHitResult.createMissed(
+                    end,
+                    Direction.getFacing(diff.x, diff.y, diff.z),
+                    new BlockPos(end)
+                ),
+                portals
+            );
+        }
+
+        // First ray trace normally
+        BlockHitResult hitResult = world.rayTrace(context);
+
+        List<Pair<Portal, Vec3d>> rayTracedPortals = withSwitchedContext(world,
```

Comment:
Why do you need to switch context for this?

Reply:
It makes sense to do so, especially on client

But on dedicated server, withSwitchedContext doesn't do anything, since servers already support multiple worlds

Reply:
In rayTracePortals() it seems does not use client.world field so switching is not needed?

Reply:
Not everything I do is specifically because code will break if I don't do it. Sometimes it's to make the code clearer or because it feels more comfortable to me. I switch the world here because it just makes more sense to me.

Reply:
If it breaks without that then ok